File: algorithms/18.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution
{
private:
    void nSum(vector<int> &nums, int n, int target, vector< vector<int> > &result, vector<int> &tmp, int start)
    {
        if ( n == 0 )
        {
            if ( target == 0 )
                result.push_back(tmp);
            return;
        }
        for ( int i = start; i < (int)nums.size(); ++i )
        {
            if ( nums[i]*n > target || nums.back()*n < target )
                break;
            tmp.push_back(nums[i]);
            nSum(nums, n-1, target-nums[i], result, tmp, i+1);
            tmp.pop_back();
            while ( i < (int)nums.size()-n && nums[i] == nums[i+1] )
                i += 1;
        }
        return;
    }
public:
    vector< vector<int> > fourSum(vector<int>& nums, int target)
    {
        vector< vector<int> > result;
        if ( nums.size() < 4 )
            return result;
        vector<int> tmp;
        sort(nums.begin(), nums.end());
        nSum(nums, 4, target, result, tmp, 0);
        return result;
    }
};

int main()
{
    return 0;
}

File: algorithms/python/18.py
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        nums.sort()
        rls, tmp = [], []
        def nSum(n, start, target):
            if n == 0:
                if target == 0: rls.append(tmp[:])
                return
            for i in range(start, len(nums)):
                if i > start and nums[i] == nums[i-1]:
                    continue
                if nums[i]*n > target or nums[-1]*n < target:
                    break
                tmp.append(nums[i])
                nSum(n-1, i+1, target-nums[i])
                tmp.pop()
        nSum(4, 0, target)
        return rls
